text/0722-align-mount-route-apis.md

```diff
@@ -0,0 +1,73 @@
+---
+Stage: Accepted
+Start Date: 2021-02-12
+Release Date: Unreleased
+Release Versions:
+  ember-source: vX.Y.Z
+  ember-data: vX.Y.Z
+Relevant Team(s): Ember.js
+RFC PR: https://github.com/emberjs/rfcs/pull/722
+---
+
+<!---
+Directions for above:
+
+Stage: Leave as is
+Start Date: Fill in with today's date, YYYY-MM-DD
+Release Date: Leave as is
+Release Versions: Leave as is
+Relevant Team(s): Fill this in with the [team(s)](README.md#relevant-teams) to which this RFC applies
+RFC PR: Fill this in with the URL for the Proposal RFC PR
+-->
+
+# Align mount and route APIs
+
+## Summary
+
+Update the Ember `mount` API to accept a function argument which defines the Engine’s routes similar to how the `route` API works today.
+
+## Motivation
+
+Some Ember applications utilize in-repo ember-engines strictly for their lazy loading and code organization features and not for their isolation features. In this context the isolation actually causes some maintenance overhead and prevents build optimization.
```

Comment:
Which kind of build optimization it has been preventing?

Reply:
The way our code is structured we have 1 or more engine per team, most teams build and test their features in a single engine for their day to day work. During development teams rarely need any other engines outside of their own so the optimization we are working on allows teams to build only the engine that they work in regularly.

When we enable this optimization some things break such as building links from one routed engine to another since only one engine is built and no other engine routes are known at this point.

We need the ability to build links between routes because we have some features which require links to be built correctly, for example when a user performs a search we want to be able to save a link to the search results for use later. If those links cannot be built correctly it becomes hard to test that the code is working as expected until a full build and test is performed.

This change allows us to move links to the top level application thereby allowing link building via `link-to` or `router.urlFor` to work as expected.